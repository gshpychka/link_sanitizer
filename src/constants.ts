export const telegramTokenParameterName = '/link-sanitizer/telegram-token';
export const messageHandlerEnvVars = {
  urlBlackList: 'URL_PARAMETERS_BLACKLIST',
  messageQueueUrl: 'MESSAGE_QUEUE_URL',
};
