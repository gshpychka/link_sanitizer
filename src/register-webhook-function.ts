// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for RegisterWebhookFunction
 */
export interface RegisterWebhookFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/register-webhook.
 */
export class RegisterWebhookFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: RegisterWebhookFunctionProps) {
    super(scope, id, {
      description: 'src/register-webhook.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs16.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/register-webhook.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}