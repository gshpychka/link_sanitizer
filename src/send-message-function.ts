// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for SendMessageFunction
 */
export interface SendMessageFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/send-message.
 */
export class SendMessageFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: SendMessageFunctionProps) {
    super(scope, id, {
      description: 'src/send-message.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/send-message.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}