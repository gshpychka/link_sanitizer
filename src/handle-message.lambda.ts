import { parse as urlparse, UrlWithParsedQuery, format as urlformat } from 'url';
import { SQSClient, SendMessageCommand, SendMessageCommandInput } from '@aws-sdk/client-sqs';
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { Message, Update } from 'node-telegram-bot-api';
import { messageHandlerEnvVars } from './constants';

const sqs = new SQSClient({});

export const getUrlParamBlacklist = (): string[] => {
  return JSON.parse(process.env[messageHandlerEnvVars.urlBlackList]!);
};


export const sendCleanedUrls = async (originalMessage: Message, cleanedUrls: string[]) => {
  try {
    const messageQueueUrl = process.env[messageHandlerEnvVars.messageQueueUrl]!;
    let messageText: string;

    if (cleanedUrls.length === 1) {
      messageText = `Ось лінк без трекінгу: ${cleanedUrls}`;
    } else {
      messageText = `Ось лінки без трекінгу: ${cleanedUrls.join(', ')}`;
    }
    const message: SendMessageCommandInput = {
      QueueUrl: messageQueueUrl,
      MessageBody: JSON.stringify({
        chatId: originalMessage.chat.id,
        text: messageText,
        props: { reply_to_message_id: originalMessage.message_id },
      }),
    };
    const command = new SendMessageCommand(message);
    await sqs.send(command);
  } catch (err) {
    console.error('Error sending URLs');
    throw err;
  }
};

export const removeTrackingIfPresent = (url: string, parameterBlacklist: string[]): string | null => {
  try {

    let urlParsed: UrlWithParsedQuery = urlparse(url, true);

    let trackingPresent = false;
    const query_params = urlParsed.query;

    for (const blacklistedParameter of parameterBlacklist) {
      if (blacklistedParameter in query_params) {
        delete query_params[blacklistedParameter];
        trackingPresent = true;
      }
    }

    if (trackingPresent) {
      urlParsed.query = query_params;
      // need to set `search` to `null` because `url.format` uses it by default
      // instead of `query`
      urlParsed.search = null;
      return urlformat(urlParsed);
    } else {
      return null;
    }
  } catch (err) {
    console.error('Error removing tracking');
    throw err;
  }

};

export const extractUrls = (message: Message): string[] | null => {
  const urls = [];
  try {
    if (message?.entities) {
      for (const entity of message.entities) {
        if (entity.type == 'url') {

          urls.push(message.text!.slice(entity.offset, entity.offset + entity.length));
        }
      }
    }
  } catch (err) {
    console.error('Error extracting URLs');
  }
  return (urls.length > 0) ? urls : null;
};

export const processMessage = async (msg: Message, parameterBlacklist: string[]) => {
  const urls = extractUrls(msg);
  if (!urls) {
    console.log('Message does not contain a URL');
    return;
  }

  const cleanUrls = urls.map(
    url => removeTrackingIfPresent(
      url, parameterBlacklist,
    ),
  ).filter((x): x is string => x !== null);

  if (cleanUrls.length > 0) {
    console.log('Sending cleaned URLs');
    await sendCleanedUrls(msg, cleanUrls);
  } else {
    console.log('URLs do not need cleaning');
  }

};


export const handler = async (
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
  try {

    console.log(event.body);

    const update: Update = JSON.parse(event.body!);
    const parameterBlacklist = getUrlParamBlacklist();

    if (update?.message) {
      await processMessage(update.message, parameterBlacklist);
    }

    return {
      isBase64Encoded: false,
      statusCode: 200,
      body: '0',
    };
  } catch (err) {
    console.error(`Error in handler: ${JSON.stringify(err)}`);
    return {
      isBase64Encoded: false,
      statusCode: 500,
      body: '0',
    };
  }

};
