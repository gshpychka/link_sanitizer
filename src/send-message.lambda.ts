import { SSMProvider } from '@aws-lambda-powertools/parameters/ssm';
import { SQSEvent } from 'aws-lambda';
import TelegramBot from 'node-telegram-bot-api';
import { envVars } from './constants';


const parametersProvider = new SSMProvider();

export const getTelegramToken = async (): Promise<string> => {
  try {
    const token = await parametersProvider.get(
      process.env[envVars.telegramTokenParameterName]!, { decrypt: true, maxAge: 60 * 60 * 24 },
    ) ?? null;
    if (!token) {
      throw new Error('Token is empty');
    } else {
      return token;
    }
  } catch (err) {
    console.error('Error getting parameter');
    throw err;
  }
};

export const handler = async (event: SQSEvent): Promise<void> => {
  const telegramToken = await getTelegramToken();
  const bot = new TelegramBot(telegramToken, { polling: false });
  for (const record of event.Records) {
    console.log(`Processing element: ${record.body}`);
    const message = JSON.parse(record.body);
    const sentMessage = await bot.sendMessage(
      message.chatId,
      message.text,
      message.props,
    );
    console.log(`Sent message ID: ${sentMessage.message_id}`);
  }
};
