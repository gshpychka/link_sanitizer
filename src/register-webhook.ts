import * as ssm from 'aws-cdk-lib/aws-ssm';
import * as cdk from 'aws-cdk-lib/core';
import * as cr from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { RegisterWebhookFunction } from './register-webhook-function';

export interface TelegramTokenProps {
  parameter: ssm.IStringParameter;
  version?: string | undefined;
}

export interface TelegramBotWebhookProps {
  telegramToken: TelegramTokenProps;
  webhookUrl: string;
}


export interface TelegramWebhookResourceProperties {
  tokenParameterName: string;
  webhookURL: string;
}


export class TelegramBotWebhook extends Construct {
  constructor(scope: Construct, id: string, props: TelegramBotWebhookProps) {
    super(scope, id);

    const tokenParameterName = props.telegramToken.version ?
      `${props.telegramToken.parameter.parameterName}:${props.telegramToken.version}`
      : props.telegramToken.parameter.parameterName;

    const handler = new RegisterWebhookFunction(this, 'Handler');

    props.telegramToken.parameter.grantRead(handler);


    const provider = new cr.Provider(this, 'Provider', { onEventHandler: handler });

    new cdk.CustomResource(this, 'Default', {
      resourceType: 'Custom::TelegramBotWebhook',
      serviceToken: provider.serviceToken,
      properties: {
        tokenParameterName: tokenParameterName,
        webhookURL: props.webhookUrl,
      } satisfies TelegramWebhookResourceProperties,
    },
    );
  }

}
