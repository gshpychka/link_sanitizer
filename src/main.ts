import { App, Stack, StackProps } from 'aws-cdk-lib';
import * as apigw from 'aws-cdk-lib/aws-apigateway';
import * as event_sources from 'aws-cdk-lib/aws-lambda-event-sources';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import { telegramTokenParameterName, messageHandlerEnvVars } from './constants';
import { HandleMessageFunction } from './handle-message-function';
import { SendMessageFunction } from './send-message-function';

interface LinkSanitizerProps extends StackProps {
}
export class LinkSanitizer extends Stack {
  private readonly telegram_token_parameter: ssm.IStringParameter;
  constructor(scope: Construct, id: string, props: LinkSanitizerProps = {}) {
    super(scope, id, props);

    this.telegram_token_parameter = ssm.StringParameter.fromSecureStringParameterAttributes(
      this, 'TelegramToken', {
        parameterName: telegramTokenParameterName,
      });

    const messageQueue = new sqs.Queue(this, 'MessageQueue');

    const messageHandler = new HandleMessageFunction(this, 'Handler', {
      environment: {
        [messageHandlerEnvVars.urlBlackList]: JSON.stringify(
          [
            'utm_source',
            'utm_medium',
            'utm_campaign',
            'utm_name',
            'utm_term',
            'utm_content',
            'gclid',
            'fbclid',
            'igshid',
            'mc_eid',
            'mc_cid',
            '_hsenc',
            '_hsmi',
            'hsCtaTracking',
            't',
          ]),
        [messageHandlerEnvVars.messageQueueUrl]: messageQueue.queueUrl,
      },
      retryAttempts: 0,
    });

    const messageSender = new SendMessageFunction(this, 'MessageSender', {
      retryAttempts: 0,
    });

    this.telegram_token_parameter.grantRead(messageSender);

    messageQueue.grantSendMessages(messageHandler);

    const queueSource = new event_sources.SqsEventSource(messageQueue,
      { batchSize: 1, maxConcurrency: 2 },
    );
    messageSender.addEventSource(queueSource);
    new apigw.LambdaRestApi(this, 'Api', { handler: messageHandler, proxy: true });
  }
}

const app = new App();

new LinkSanitizer(app, 'LinkSanitizer');

app.synth();
