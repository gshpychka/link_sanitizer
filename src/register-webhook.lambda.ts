import crypto from 'crypto';
import { SSMProvider } from '@aws-lambda-powertools/parameters/ssm';
import { CdkCustomResourceResponse } from 'aws-lambda';
import TelegramBot from 'node-telegram-bot-api';
import { TelegramWebhookResourceProperties } from './register-webhook';


const parametersProvider = new SSMProvider();

export const getTelegramToken = async (parameterName: string): Promise<string> => {
  try {
    const token = await parametersProvider.get(
      parameterName, { decrypt: true, maxAge: 60 * 60 * 24 },
    ) ?? null;
    if (!token) {
      throw new Error('Token is empty');
    } else {
      return token;
    }
  } catch (err) {
    console.error('Error getting parameter');
    throw err;
  }
};

interface CustomResourceEvent {
  ResourceProperties: TelegramWebhookResourceProperties;
  RequestType: 'Create' | 'Update' | 'Delete';
}


export async function handler(event: CustomResourceEvent): Promise<CdkCustomResourceResponse> {
  console.log('Received event:', JSON.stringify(event, null, 2));

  const properties: TelegramWebhookResourceProperties = event.ResourceProperties;

  try {
    const token = await getTelegramToken(properties.tokenParameterName);
    const physicalId = crypto.createHash('sha256').update(token).digest('hex');

    const bot = new TelegramBot(token);

    let response: any;

    if (event.RequestType === 'Delete') {
      response = await bot.deleteWebHook();
    } else {
      response = await bot.setWebHook(properties.webhookURL);
    }

    if (!response) {
      throw new Error(`Failed to ${event.RequestType.toLowerCase()} webhook`);
    }

    return {
      PhysicalResourceId: physicalId,
    };
  } catch (error) {
    console.error('Error:', error);
    throw error;
  }
}
