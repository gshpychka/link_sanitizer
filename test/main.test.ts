import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { envVars } from '../src/constants';
import { LinkSanitizer } from '../src/main';

test('LinkSanitizer stack', () => {
  const app = new App();
  const tokenParameterName = 'MyToken';
  const stack = new LinkSanitizer(app, 'test', { telegramTokenParameterName: tokenParameterName });

  const template = Template.fromStack(stack);

  const messageHandlerId = 'Handler886CB40B';
  const messageSenderId = 'MessageSenderC8564C19';
  const queueId = 'MessageQueue7A3BF959';

  // Check the snapshot
  expect(template.toJSON()).toMatchSnapshot();

  // Check there is one queue created
  const queues = template.findResources('AWS::SQS::Queue', {});

  expect(Object.keys(queues)).toHaveLength(1);

  expect(queues).toHaveProperty(queueId);

  // Check there are two Lambda functions created
  const lambdas = template.findResources('AWS::Lambda::Function', {});
  expect(Object.keys(lambdas)).toHaveLength(2);

  const messageHandlerLambda = lambdas[messageHandlerId];
  expect(messageHandlerLambda).toBeDefined();

  const messageSenderLambda = lambdas[messageSenderId];
  expect(messageSenderLambda).toBeDefined();

  // Check URL_BLACK_LIST is a JSON array of strings
  const urlBlackList = JSON.parse(
    messageHandlerLambda.Properties.Environment.Variables[envVars.urlBlackList],
  );
  expect(Array.isArray(urlBlackList)).toBe(true);
  urlBlackList.forEach((item: string) => {
    expect(typeof item).toBe('string');
  });

  expect(
    messageHandlerLambda.Properties.Environment.Variables[
      envVars.messageQueueUrl
    ],
  ).toEqual({ Ref: queueId });

  expect(
    messageSenderLambda.Properties.Environment.Variables[
      envVars.telegramTokenParameterName
    ],
  ).toEqual(tokenParameterName);

  const eventSourceMappings = template.findResources('AWS::Lambda::EventSourceMapping', {
    Properties: {
      EventSourceArn: { 'Fn::GetAtt': [queueId, 'Arn'] },
      FunctionName: { Ref: messageSenderId },
    },
  });
  expect(Object.keys(eventSourceMappings)).toHaveLength(1);


  const messageHandlerRolePolicies = template.findResources('AWS::IAM::Policy', {
    Properties: {
      Roles: [
        {
          Ref: messageHandlerLambda.Properties.Role['Fn::GetAtt'][0],
        },
      ],
    },
  });

  expect(Object.keys(messageHandlerRolePolicies)).toHaveLength(1);

  interface PolicyStatement {
    Action: string | string[];
    Resource: { 'Fn::GetAtt': [string, string] };
  }


  const handlerHasWriteAccessToQueue = Object.values(messageHandlerRolePolicies).some(
    policy => {
      return policy.Properties.PolicyDocument.Statement.some((statement: PolicyStatement) => {
        const actions = Array.isArray(statement.Action) ? statement.Action : [statement.Action];
        return actions.includes('sqs:SendMessage') &&
          statement.Resource['Fn::GetAtt'][0] === queueId;
      },
      );
    });

  expect(handlerHasWriteAccessToQueue).toBe(true);


  const messageSenderRolePolicies = template.findResources('AWS::IAM::Policy', {
    Properties: {
      Roles: [
        {
          Ref: messageSenderLambda.Properties.Role['Fn::GetAtt'][0],
        },
      ],
    },
  });
  const senderHasReadAccessToQueue = Object.values(messageSenderRolePolicies).some(
    policy => {
      return policy.Properties.PolicyDocument.Statement.some((statement: PolicyStatement) => {
        const actions = Array.isArray(statement.Action) ? statement.Action : [statement.Action];
        return actions.includes('sqs:ReceiveMessage') &&
          statement.Resource['Fn::GetAtt'][0] === queueId;
      },
      );
    });

  expect(senderHasReadAccessToQueue).toBe(true);

});
