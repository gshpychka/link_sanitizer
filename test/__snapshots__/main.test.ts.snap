// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkSanitizer stack 1`] = `
{
  "Outputs": {
    "ApiEndpoint4F160690": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "ApiF70053CD",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "ApiDeploymentStageprod3EB9684E",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiANY40671C4F": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "Handler886CB40B",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiANYApiPermissionTesttestApi91C49FB9ANYF2F064A4": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiF70053CD",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiANYApiPermissiontestApi91C49FB9ANY1F9E02F6": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiF70053CD",
              },
              "/",
              {
                "Ref": "ApiDeploymentStageprod3EB9684E",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiAccountA18C9B29": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "ApiF70053CD",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "ApiCloudWatchRole73EC6FC4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiCloudWatchRole73EC6FC4": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiDeploymentB17BE62D6b6eb4522b196cdc7aa3bb6ccfa03df7": {
      "DependsOn": [
        "ApiproxyANY4946669B",
        "Apiproxy05443892",
        "ApiANY40671C4F",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiDeploymentStageprod3EB9684E": {
      "DependsOn": [
        "ApiAccountA18C9B29",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "ApiDeploymentB17BE62D6b6eb4522b196cdc7aa3bb6ccfa03df7",
        },
        "RestApiId": {
          "Ref": "ApiF70053CD",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiF70053CD": {
      "Properties": {
        "Name": "Api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "Apiproxy05443892": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ApiproxyANY4946669B": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "Handler886CB40B",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "Apiproxy05443892",
        },
        "RestApiId": {
          "Ref": "ApiF70053CD",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiproxyANYApiPermissionTesttestApi91C49FB9ANYproxy7A023D0A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiF70053CD",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiproxyANYApiPermissiontestApi91C49FB9ANYproxy19423C2D": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiF70053CD",
              },
              "/",
              {
                "Ref": "ApiDeploymentStageprod3EB9684E",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "Handler886CB40B": {
      "DependsOn": [
        "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "HandlerServiceRoleFCDC14AE",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "24de3b1d7721e0e621df7f9128feabe1de7cf307c7cb03a0d3e78acc63fbdeea.zip",
        },
        "Description": "src/handle-message.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "MESSAGE_QUEUE_URL": {
              "Ref": "MessageQueue7A3BF959",
            },
            "URL_PARAMETERS_BLACKLIST": "["utm_source","utm_medium","utm_campaign","utm_name","utm_term","utm_content","gclid","fbclid","igshid","mc_eid","mc_cid","_hsenc","_hsmi","hsCtaTracking","t"]",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "HandlerServiceRoleFCDC14AE",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "HandlerEventInvokeConfig16376DFA": {
      "Properties": {
        "FunctionName": {
          "Ref": "Handler886CB40B",
        },
        "MaximumRetryAttempts": 0,
        "Qualifier": "$LATEST",
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "HandlerServiceRoleDefaultPolicyCBD0CC91": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MessageQueue7A3BF959",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "Roles": [
          {
            "Ref": "HandlerServiceRoleFCDC14AE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HandlerServiceRoleFCDC14AE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MessageQueue7A3BF959": {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "MessageSenderC8564C19": {
      "DependsOn": [
        "MessageSenderServiceRoleDefaultPolicy568938A7",
        "MessageSenderServiceRole0B8B9CA7",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "3398dc73bff86166d0bd8466f79c7cbf58e53e518db280e4f9b66e47d4347673.zip",
        },
        "Description": "src/send-message.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "MessageSenderServiceRole0B8B9CA7",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MessageSenderEventInvokeConfig463CA2E9": {
      "Properties": {
        "FunctionName": {
          "Ref": "MessageSenderC8564C19",
        },
        "MaximumRetryAttempts": 0,
        "Qualifier": "$LATEST",
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "MessageSenderServiceRole0B8B9CA7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MessageSenderServiceRoleDefaultPolicy568938A7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/link-sanitizer/telegram-token",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MessageQueue7A3BF959",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MessageSenderServiceRoleDefaultPolicy568938A7",
        "Roles": [
          {
            "Ref": "MessageSenderServiceRole0B8B9CA7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MessageSenderSqsEventSourcetestMessageQueue7CFBF9F0387EAD69": {
      "Properties": {
        "BatchSize": 1,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "MessageQueue7A3BF959",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "MessageSenderC8564C19",
        },
        "ScalingConfig": {
          "MaximumConcurrency": 2,
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
